core/includes/backup/backup.database.inc

```diff
@@ -0,0 +1,350 @@
+      // This ability is not used directly by core, which only provides UIs for
+      // backing up and restoring existing database connections.
+      else {
+        $url_parts = parse_url($this->getTarget());
+        // If the url is specified build it into a connection info array.
+        if ($url_parts) {
+          $info = array(
+            'driver' => empty($url_parts['scheme']) ? NULL : $url_parts['scheme'],
+            'host' => empty($url_parts['host']) ? NULL : $url_parts['host'],
+            'port' => empty($url_parts['port']) ? 3306 : $url_parts['port'],
+            'username' => empty($url_parts['user']) ? NULL : $url_parts['user'],
+            'password' => empty($url_parts['pass']) ? NULL : $url_parts['pass'],
+            'database' => empty($url_parts['path']) ? NULL : $url_parts['path'],
+          );
+          $key = uniqid('backdrop_backup_tmp_');
+          $target = 'default';
+          Database::addConnectionInfo($key, $target, $info);
+        }
+        // No database selected. Assume the default.
+        else {
+          $key = $target = 'default';
+        }
+      }
+      if ($target && $key) {
+        $this->connection = Database::getConnection($target, $key);
+      }
+    }
+    return $this->connection;
+  }
+
+  /**
+   * Backup the databases to a file.
+   *
+   * @return bool
+   *   TRUE is backup successful, FALSE otherwise.
+   */
+  abstract protected function backupDatabaseToFile(BackupFile $file);
+
+  /**
+   * Backup the databases to a file.
+   *
+   * @return int|false
+   *   The number of commands executed on restore. FALSE on failure.
+   */
+  abstract protected function restoreDatabaseFromFile(BackupFile $file);
+
+  /**
+   * Get a list of objects in the database.
+   */
+  protected function getObjectNames() {
+    // Must be overridden.
+    $out = $this->getTableNames();
+    if (method_exists($this, 'getViewNames')) {
+      $out += $this->getViewNames();
+    }
+    return $out;
+  }
+
+  /**
+   * Get a list of tables in the database.
+   *
+   * @return string[]
+   *   An array of all table names keyed by the table name.
+   */
+  abstract protected function getTableNames();
+
+  /**
+   * Get a list of views in the database.
+   *
+   * @return string[]
+   *   An array of all view names keyed by the view name.
+   */
+  abstract protected function getViewNames();
+
+  /**
+   * Get a list of tables to be locked during the backup.
+   *
+   * @return array
+   *   The list of tables to be locked, if any.
+   */
+  protected function getLockedTables() {
+    $tables = array();
+    if ($this->settings['lock_tables']) {
+      foreach ($this->getTableNames() as $table) {
+        // There's no need to lock excluded or structure only tables because it
+        // doesn't matter if they change.
+        if (empty($settings['exclude_tables']) || !in_array($table, (array) $settings['exclude_tables'])) {
```

Comment:
Variable `$settings` is never defined here. Shouldn't that rather be `$this->settings`? Maybe that's why this is casted to array, to prevent PHP nagging.

But I wonder, if we need that exclude_tables setting, anyway. Where would it be set?

Reply:
Thanks, nice catch.

> But I wonder, if we need that exclude_tables setting, anyway. Where would it be set?

It's set in `BackupDatabase::defaultSettings()`. It's possible to override via the API, but the defaults exclude all `simpletest_*` tables (those used when running tests).